python/deltalake/table.py

```diff
@@ -1435,3 +1435,18 @@ def z_order(
         )
         self.table.update_incremental()
         return json.loads(metrics)
+
+
+def _cast_to_equal_batch(
+    batch: pyarrow.RecordBatch, schema: pyarrow.Schema
+) -> pyarrow.RecordBatch:
+    """
+    Cast a batch to a schema, if it is already considered equal.
+
+    This is mostly for mapping things like list field names, which arrow-rs
+    checks when looking at schema equality, but pyarrow does not.
+    """
+    if batch.schema == schema:
+        return pyarrow.Table.from_batches([batch]).cast(schema).to_batches()[0]
+    else:
+        return batch
```

Comment:
@ion-elgreco If we take this out, do the tests still pass?

Reply:
Yes, everything passes still. Shall I take it out? 

Or would we require it later once we enable the compliant types again.

Reply:
Ideally I think we shouldn’t need it. This was an earlier attempt at fixing the compliant types thing

Reply:
We should have backwards compatibility during reads and writes when we use the compliant types, I think spark-delta also doesn't write these compliant types but I may be wrong here